src/utils/allocators.zig

```diff
@@ -178,120 +192,154 @@ pub const DiskMemoryAllocator = struct {
+            std.mem.bytesAsValue(Metadata, buf_ptr[new_size..][0..@sizeOf(Metadata)]).* = .{
+                .file_index = metadata.file_index,
+            };
+            return true;
+        } else {
+            const file = self.dir.openFile(file_name, .{ .mode = .read_write }) catch |err| {
+                self.logFailure(err, file_name);
+                return false;
+            };
+            defer file.close();
+
+            const mapped = std.posix.mmap(
+                buf_ptr,
+                new_aligned_size,
+                std.posix.PROT.READ | std.posix.PROT.WRITE,
+                std.posix.MAP{ .TYPE = .SHARED },
+                file.handle,
+                0,
+            ) catch |err| {
+                self.logFailure(err, file_name);
+                return false;
+            };
+            std.debug.assert(mapped.ptr == buf_ptr);
+
+            return true;
+        }
     }
-};
 
-test "recycle allocator" {
-    const backing_allocator = std.testing.allocator;
-    var allocator = try RecycleFBA(.{}).init(backing_allocator, 1024);
-    defer allocator.deinit();
+    /// unmaps the memory (file still exists and is removed on deinit())
+    fn free(ctx: *anyopaque, buf: []u8, log2_align: u8, return_address: usize) void {
+        _ = return_address;
+        const self: *Self = @ptrCast(@alignCast(ctx));
 
-    // alloc a slice of 100 bytes
-    const bytes = try allocator.allocator().alloc(u8, 100);
-    const ptr = bytes.ptr;
-    // free the slice
-    allocator.allocator().free(bytes);
+        const alignment = @as(usize, 1) << @intCast(log2_align);
+        std.debug.assert(alignment <= std.mem.page_size); // the allocator interface shouldn't allow this (aside from the *Raw methods).
 
-    // realloc should be the same (ie, recycled data)
-    const bytes2 = try allocator.allocator().alloc(u8, 100);
-    try std.testing.expectEqual(ptr, bytes2.ptr);
-    allocator.allocator().free(bytes2);
+        const buf_ptr: [*]align(std.mem.page_size) u8 = @alignCast(buf.ptr);
+        const aligned_size = fullMmapSize(buf.len);
 
-    // same result with smaller slice
-    const bytes3 = try allocator.allocator().alloc(u8, 50);
-    try std.testing.expectEqual(ptr, bytes3.ptr);
-    allocator.allocator().free(bytes3);
+        const metadata: Metadata = @bitCast(buf_ptr[buf.len..][0..@sizeOf(Metadata)].*);
+        const file_name_bounded = fileNameBounded(metadata.file_index);
+        const file_name = file_name_bounded.constSlice();
 
-    // diff result with larger slice
-    const bytes4 = try allocator.allocator().alloc(u8, 200);
-    try std.testing.expect(ptr != bytes4.ptr);
-    allocator.allocator().free(bytes4);
-}
+        std.posix.munmap(buf_ptr[0..aligned_size]);
+        self.dir.deleteFile(file_name) catch |err| {
+            self.logFailure(err, file_name);
+        };
+    }
+
+    const Metadata = extern struct {
```

Comment:
can you define this underneath the fields -- also can u update the contributing .md to describe this 

we want it ordered like `struct { fields, self = @This(), structs, init, deinit, methods }`

Reply:
that makes sense - maybe we can add a comment how its a general rule and isnt strict - adding the example you mentioned

Reply:
The main reason I put the Metadata struct where it is is because it's more of an implementation detail to the allocator, there's no real way for a user of this struct to interface with this part of the DMA. I'm fine to move it up if you're set on that, but imo it's mostly clutter as opposed to actually useful information.

Reply:
talked about this offline - for consistency with the rest of the codebase were gonna move Metadata to the top

Reply:
6c5d2322b3903eb4e646722255223fe931a3e14c